themes/api.py

```diff
@@ -0,0 +1,135 @@
+"""Themes APIs"""
+import praw
+
+from django.conf import settings
+
+THEME_TYPE_PUBLIC = 'public'
+THEME_TYPE_PRIVATE = 'private'
+
+VALID_THEME_TYPES = (
+    THEME_TYPE_PRIVATE,
+    THEME_TYPE_PUBLIC,
+)
+
+USER_AGENT = 'MIT-Open: {version}'
+
+
+def _get_user_credentials(user):
+    """
+    Get credentials for authenticated user
+
+    Args:
+        user (User): the authenticated user
+
+    Returns:
+        dict: set of configuration credentials for the user
+    """
+    # pylint: disable=fixme,unused-argument
+    # TODO: get a refresh token
+    return {
+        'client_id': settings.MIT_OPEN_REDDIT_AUTHENTICATED_CLIENT_ID,
+        'client_secret': settings.MIT_OPEN_REDDIT_AUTHENTICATED_SECRET,
+    }
+
+
+def _get_anonymous_credentials():
+    """
+    Get credentials for anonymous user
+
+    Returns:
+        dict: set of configuration credentials for the user
+    """
+    return {
+        'client_id': settings.MIT_OPEN_REDDIT_ANONYMOUS_CLIENT_ID,
+        'client_secret': settings.MIT_OPEN_REDDIT_ANONYMOUS_SECRET,
+        'username': settings.MIT_OPEN_REDDIT_ANONYMOUS_USERNAME,
+        'password': settings.MIT_OPEN_REDDIT_ANONYMOUS_PASSWORD,
+    }
+
+
+def _get_credentials(user=None):
+    """
+    Get credentials for authenticated or anonymous user
+
+    Args:
+        user (User): the authenticated user
+
+    Returns:
+        dict: set of configuration credentials for the user
+    """
+    return _get_user_credentials(user) if user is not None else _get_anonymous_credentials()
+
+
+def _get_client(user=None):
+    """
+    Get a configured Reddit client_id
+
+    Args:
+        user (User): the authenticated user
+
+    Returns:
+        praw.Reddit: configured reddit client
+    """
+    credentials = _get_credentials(user=user)
+
+    return praw.Reddit(
+        reddit_url=settings.MIT_OPEN_REDDIT_URL,
+        oauth_url=settings.MIT_OPEN_REDDIT_URL,
+        short_url=settings.MIT_OPEN_REDDIT_URL,
+        user_agent=_get_user_agent(),
+        **credentials
+    )
+
+
+def _get_user_agent():
+    """Gets the user agent"""
+    return USER_AGENT.format(version=settings.VERSION)
+
+
+class Api:
+    """Theme API"""
+    def __init__(self, user=None):
+        """Constructor"""
+        self.user = user
+        self.reddit = _get_client(user=user)
+
+    @property
+    def is_anonymous(self):
+        """Returns True if user is anonymous"""
+        return self.user is None
+
+    def list_themes(self):
+        """
+        List the themes
+
+        Returns:
+            ListingGenerator(praw.models.Subreddit): a generator over theme listings
+        """
+        return self.reddit.user.subreddits() if self.is_anonymous is not None else self.reddit.subreddits.default()
+
+    def create_theme(self, name, title, theme_type=THEME_TYPE_PUBLIC):
```

Comment:
Ok, the neither the `praw` documentation or reddit's was clear on that, but in testing the api accepts those options. I'll implement `**other_settings` with at least some light validation around it since it's a fixed set of args.